File: app/src/main/java/ru/magflayer/spectrum/domain/manager/CameraManager.java
package ru.magflayer.spectrum.domain.manager;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.ImageFormat;
import android.graphics.Rect;
import android.graphics.SurfaceTexture;
import android.graphics.YuvImage;
import android.hardware.Camera;
import android.support.annotation.Nullable;
import android.view.Display;
import android.view.Surface;
import android.view.WindowManager;

import com.squareup.otto.Bus;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

import javax.inject.Inject;
import javax.inject.Singleton;

import ru.magflayer.spectrum.common.utils.RxUtils;
import ru.magflayer.spectrum.presentation.common.utils.BitmapUtils;
import rx.Observable;
import rx.Subscription;

@Singleton
@SuppressWarnings({"deprecation", "WeakerAccess"})
public class CameraManager {

    private final Logger log = LoggerFactory.getLogger(getClass());

    private static final int CAMERA_WIDTH = 1280;
    private static final int CAMERA_HEIGHT = 720;

    interface OnTakePictureListener {
        void onTakePicture(Bitmap bitmap);
    }

    @Inject
    Bus bus;

    @Nullable
    private Camera camera;
    private Bitmap cameraBitmap;
    private Subscription generateBitmapSubscription;
    private ByteArrayOutputStream generateBitmapStream;
    private WindowManager windowManager;

    @Inject
    CameraManager(final Context context) {
        windowManager = ((WindowManager) context.getSystemService(Context.WINDOW_SERVICE));
    }

    public void open() {
        camera = Camera.open(getBackFacingCameraId());

        if (camera == null) {
            throw new RuntimeException("Camera not available");
        }

        if (generateBitmapStream == null) {
            generateBitmapStream = new ByteArrayOutputStream();
        }

        Camera.Parameters parameters = camera.getParameters();
        parameters.setPreviewSize(CAMERA_WIDTH, CAMERA_HEIGHT);
        parameters.setPreviewFormat(ImageFormat.NV21);
        camera.setParameters(parameters);
        camera.setPreviewCallback((data, camera1) -> {
            if (generateBitmapSubscription == null || generateBitmapSubscription.isUnsubscribed()) {
                Camera.Size size = camera.getParameters().getPreviewSize();
                generateBitmapSubscription = generateBitmapObservable(data, size, generateBitmapStream)
                        .subscribe(bitmap -> cameraBitmap = bitmap,
                                throwable -> log.error("Error while generating bitmap: ", throwable));
            }
        });
    }

    public void startCamera(@Nullable SurfaceTexture texture) throws IOException {
        if (texture == null) {
            throw new NullPointerException("You cannot start preview without a preview texture");
        }

        if (camera == null) {
            throw new IllegalStateException("You should call open() method before start camera");
        }

        camera.setPreviewTexture(texture);
        camera.startPreview();

        cameraBitmap = Bitmap.createBitmap(CAMERA_WIDTH, CAMERA_HEIGHT, Bitmap.Config.ARGB_8888);
    }

    public void autoFocus() {
        if (camera != null) {
            try {
                camera.autoFocus((success, camera1) -> camera1.cancelAutoFocus());
            } catch (RuntimeException e) {
                log.warn("Cannot auto focus camera: ", e);
            }
        }
    }

    public void takePicture(final OnTakePictureListener pictureCallback) {
        if (camera != null) {
            camera.takePicture(shutterCallback, rawCallback, (data, camera1) -> {
                log.debug("onPictureTaken - jpeg");

                Bitmap bitmap = BitmapUtils.bytesToBitmap(data);
                pictureCallback.onTakePicture(bitmap);

                camera1.startPreview();
            });
        }
    }

    public Bitmap getCameraBitmap() {
        return cameraBitmap;
    }

    public void stopPreview() {
        if (camera != null) {
            camera.stopPreview();
        }
    }

    public void close() {
        if (camera != null) {
            camera.stopPreview();
            camera.setPreviewCallback(null);
            camera.unlock();
            camera.release();
            camera = null;
        }
        if (generateBitmapSubscription != null) {
            generateBitmapSubscription.unsubscribe();
            generateBitmapSubscription = null;
        }
        if (generateBitmapStream != null) {
            try {
                generateBitmapStream.close();
            } catch (IOException e) {
                log.warn("Error occurred while closing generate bitmap stream: ", e);
            }
            generateBitmapStream = null;
        }
    }

    public void updateCameraDisplayOrientation() {
        android.hardware.Camera.CameraInfo camInfo =
                new android.hardware.Camera.CameraInfo();
        android.hardware.Camera.getCameraInfo(getBackFacingCameraId(), camInfo);

        Display display = windowManager.getDefaultDisplay();
        int rotation = display.getRotation();
        int degrees = 0;
        switch (rotation) {
            case Surface.ROTATION_0:
                degrees = 0;
                break;
            case Surface.ROTATION_90:
                degrees = 90;
                break;
            case Surface.ROTATION_180:
                degrees = 180;
                break;
            case Surface.ROTATION_270:
                degrees = 270;
                break;
        }

        updateCameraDisplayOrientation(degrees);
    }

    public void updateCameraDisplayOrientation(final int degrees) {
        android.hardware.Camera.CameraInfo camInfo =
                new android.hardware.Camera.CameraInfo();
        android.hardware.Camera.getCameraInfo(getBackFacingCameraId(), camInfo);

        int result;
        if (camInfo.facing == Camera.CameraInfo.CAMERA_FACING_FRONT) {
            result = (camInfo.orientation + degrees) % 360;
            result = (360 - result) % 360;  // compensate the mirror
        } else {  // back-facing
            result = (camInfo.orientation - degrees + 360) % 360;
        }

        if (camera != null) {
            camera.setDisplayOrientation(result);
        }
    }

    private int getBackFacingCameraId() {
        int cameraId = -1;

        int numberOfCameras = Camera.getNumberOfCameras();
        for (int i = 0; i < numberOfCameras; i++) {
            Camera.CameraInfo info = new Camera.CameraInfo();
            Camera.getCameraInfo(i, info);
            if (info.facing == Camera.CameraInfo.CAMERA_FACING_BACK) {
                cameraId = i;
                break;
            }
        }
        return cameraId;
    }

    private Camera.ShutterCallback shutterCallback = () -> log.debug("onShutter");

    private Camera.PictureCallback rawCallback = (data, camera1) -> log.debug("onPictureTaken - raw");

    private Observable<Bitmap> generateBitmapObservable(final byte[] data,
                                                        final Camera.Size previewSize,
                                                        final ByteArrayOutputStream stream) {
        return Observable.just(data)
                .compose(RxUtils.applySchedulers(RxUtils.ANDROID_THREAD_POOL_EXECUTOR, RxUtils.ANDROID_THREAD_POOL_EXECUTOR))
                .flatMap(bytes -> {
                    //Convert YUV to RGB
                    YuvImage yuvimage = new YuvImage(data, ImageFormat.NV21, previewSize.width, previewSize.height, null);
                    yuvimage.compressToJpeg(new Rect(0, 0, previewSize.width, previewSize.height), 50, stream);
                    byte[] jdata = stream.toByteArray();
                    stream.reset();
                    return Observable.just(BitmapFactory.decodeByteArray(jdata, 0, jdata.length));

                    //TODO research decode Yuv data to RGB, below method very slow
//                      int[] imagePixels = convertYUV420_NV21toRGB8888(yuvimage.getYuvData(), previewSize.width, previewSize.height);
//                      return Observable.just(Bitmap.createBitmap(imagePixels, previewSize.width, previewSize.height, Bitmap.Config.ARGB_8888));
                });
    }

    /**
     * Converts YUV420 NV21 to RGB8888
     *
     * @param data   byte array on YUV420 NV21 format.
     * @param width  pixels width
     * @param height pixels height
     * @return a RGB8888 pixels int array. Where each int is a pixels ARGB.
     */
    @SuppressWarnings("UnnecessaryLocalVariable")
    public static int[] convertYUV420_NV21toRGB8888(byte[] data, int width, int height) {
        int size = width * height;
        int offset = size;
        int[] pixels = new int[size];
        int u, v, y1, y2, y3, y4;

        // i percorre os Y and the final pixels
        // k percorre os pixles U e V
        for (int i = 0, k = 0; i < size; i += 2, k += 2) {
            y1 = data[i] & 0xff;
            y2 = data[i + 1] & 0xff;
            y3 = data[width + i] & 0xff;
            y4 = data[width + i + 1] & 0xff;

            u = data[offset + k] & 0xff;
            v = data[offset + k + 1] & 0xff;
            u = u - 128;
            v = v - 128;

            pixels[i] = convertYUVtoRGB(y1, u, v);
            pixels[i + 1] = convertYUVtoRGB(y2, u, v);
            pixels[width + i] = convertYUVtoRGB(y3, u, v);
            pixels[width + i + 1] = convertYUVtoRGB(y4, u, v);

            if (i != 0 && (i + 2) % width == 0)
                i += width;
        }

        return pixels;
    }

    private static int convertYUVtoRGB(int y, int u, int v) {
        int r, g, b;

        r = y + (int) (1.402f * v);
        g = y - (int) (0.344f * u + 0.714f * v);
        b = y + (int) (1.772f * u);
        r = r > 255 ? 255 : r < 0 ? 0 : r;
        g = g > 255 ? 255 : g < 0 ? 0 : g;
        b = b > 255 ? 255 : b < 0 ? 0 : b;
        return 0xff000000 | (b << 16) | (g << 8) | r;
    }
}

File: app/src/main/java/ru/magflayer/spectrum/domain/manager/CameraManager.kt
package ru.magflayer.spectrum.domain.manager

import android.content.Context
import android.content.pm.PackageManager
import android.graphics.*
import android.hardware.Camera
import android.view.Surface
import android.view.WindowManager
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.*
import org.slf4j.LoggerFactory
import java.io.ByteArrayOutputStream
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.experimental.and

@Suppress("DEPRECATION")
@Singleton
class CameraManager @Inject constructor(
    @ApplicationContext val context: Context
) {

    private val log = LoggerFactory.getLogger(javaClass)

    private var camera: Camera? = null
    var cameraBitmap: Bitmap? = null
        private set
    private var cameraScope: CoroutineScope? = null
    private val generateBitmapStream = ByteArrayOutputStream()
    private val windowManager: WindowManager =
        context.getSystemService(Context.WINDOW_SERVICE) as WindowManager

    private val backFacingCameraId: Int
        get() {
            var cameraId = -1

            val numberOfCameras = Camera.getNumberOfCameras()
            for (i in 0 until numberOfCameras) {
                val info = Camera.CameraInfo()
                Camera.getCameraInfo(i, info)
                if (info.facing == Camera.CameraInfo.CAMERA_FACING_BACK) {
                    cameraId = i
                    break
                }
            }
            return cameraId
        }

    private val shutterCallback = Camera.ShutterCallback { log.debug("onShutter") }
    private val rawCallback = Camera.PictureCallback { _, _ -> log.debug("onPictureTaken - raw") }

    interface OnTakePictureListener {
        fun onTakePicture(bitmap: Bitmap)
    }

    fun isFlashAvailable() =
        context.packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_FLASH)

    fun open() {
        camera = Camera.open(backFacingCameraId)
        cameraScope = CoroutineScope(Dispatchers.Default)

        if (camera == null) {
            throw RuntimeException("Camera not available")
        }

        camera?.apply {
            val params = parameters
            params.setPreviewSize(CAMERA_WIDTH, CAMERA_HEIGHT)
            params.previewFormat = ImageFormat.NV21
            parameters = params
        }

        camera?.setPreviewCallback { data, _ ->
            val errorHandler = CoroutineExceptionHandler { _, exception ->
                log.warn("Error while generating bitmap: ", exception)
            }
            cameraScope?.launch(errorHandler) {
                camera?.parameters?.previewSize?.let { size ->
                    cameraBitmap = generateBitmapObservable(data, size, generateBitmapStream)
                }
            }
        }
    }

    @Throws(IOException::class)
    fun startCamera(texture: SurfaceTexture?) {
        if (texture == null) {
            throw NullPointerException("You cannot start preview without a preview texture")
        }

        if (camera == null) {
            throw IllegalStateException("You should call open() method before start camera")
        }

        camera?.setPreviewTexture(texture)
        camera?.startPreview()

        cameraBitmap = Bitmap.createBitmap(CAMERA_WIDTH, CAMERA_HEIGHT, Bitmap.Config.ARGB_8888)
    }

    fun autoFocus() {
        try {
            camera?.autoFocus { _, camera1 -> camera1.cancelAutoFocus() }
        } catch (e: RuntimeException) {
            log.warn("Cannot auto focus camera: ", e)
        }
    }

    fun isFlashlightEnabled(): Boolean {
        return camera?.run {
            parameters.flashMode == Camera.Parameters.FLASH_MODE_TORCH
                    || parameters.flashMode == Camera.Parameters.FLASH_MODE_ON
        } ?: false
    }

    fun enabledFlash() {
        try {
            camera?.let {
                val params = it.parameters
                params.flashMode = Camera.Parameters.FLASH_MODE_TORCH
                it.parameters = params
            }
        } catch (e: Exception) {
            log.warn("Cannot set parameters: ", e)
        }
    }

    fun disableFlash() {
        try {
            camera?.let {
                val params = it.parameters
                params.flashMode = Camera.Parameters.FLASH_MODE_OFF
                it.parameters = params
            }
        } catch (e: Exception) {
            log.warn("Cannot set parameters: ", e)
        }
    }

    fun isZoomSupported(): Boolean {
        return camera?.parameters?.isZoomSupported ?: false
    }

    fun getZoom(): Int {
        return camera?.parameters?.zoom ?: 0
    }

    fun getZoomRatio(): Int {
        return camera?.run {
            val index = parameters.zoom
            parameters.zoomRatios?.get(index)
        } ?: 100
    }

    fun getMaxZoom(): Int {
        return camera?.parameters?.maxZoom ?: 0
    }

    fun setZoom(zoom: Int) {
        camera?.cancelAutoFocus()
        camera?.apply {
            val param = parameters
            param.zoom = zoom
            parameters = param
        }
    }

    fun takePicture(pictureCallback: OnTakePictureListener) {
        camera?.takePicture(shutterCallback, rawCallback, { data, camera1 ->
            log.debug("onPictureTaken - jpeg")

            val bitmap = BitmapFactory.decodeByteArray(data, 0, data.size)
            pictureCallback.onTakePicture(bitmap)

            camera1.startPreview()
        })
    }

    fun stopPreview() {
        camera?.stopPreview()
    }

    fun close() {
        camera?.apply {
            stopPreview()
            setPreviewCallback(null)
            unlock()
            release()
        }
        camera = null
        cameraScope?.cancel()
        generateBitmapStream.reset()
    }

    fun updateCameraDisplayOrientation() {
        val camInfo = Camera.CameraInfo()
        Camera.getCameraInfo(backFacingCameraId, camInfo)

        val display = windowManager.defaultDisplay
        val rotation = display.rotation
        var degrees = 0
        when (rotation) {
            Surface.ROTATION_0 -> degrees = 0
            Surface.ROTATION_90 -> degrees = 90
            Surface.ROTATION_180 -> degrees = 180
            Surface.ROTATION_270 -> degrees = 270
        }

        updateCameraDisplayOrientation(degrees)
    }

    private fun updateCameraDisplayOrientation(degrees: Int) {
        val camInfo = Camera.CameraInfo()
        Camera.getCameraInfo(backFacingCameraId, camInfo)

        var result: Int
        if (camInfo.facing == Camera.CameraInfo.CAMERA_FACING_FRONT) {
            result = (camInfo.orientation + degrees) % 360
            result = (360 - result) % 360  // compensate the mirror
        } else {  // back-facing
            result = (camInfo.orientation - degrees + 360) % 360
        }

        camera?.setDisplayOrientation(result)
    }

    private fun generateBitmapObservable(
        data: ByteArray,
        previewSize: Camera.Size,
        stream: ByteArrayOutputStream
    ): Bitmap {
        //Convert YUV to RGB
        val yuvImage = YuvImage(data, ImageFormat.NV21, previewSize.width, previewSize.height, null)
        yuvImage.compressToJpeg(
            Rect(0, 0, previewSize.width, previewSize.height),
            50,
            stream
        )
        val dataBytes = stream.toByteArray()
        stream.reset()
        return BitmapFactory.decodeByteArray(dataBytes, 0, dataBytes.size)
            ?: Bitmap.createBitmap(CAMERA_WIDTH, CAMERA_HEIGHT, Bitmap.Config.ARGB_8888)

        //TODO research decode Yuv data to RGB, below method very slow
        // int[] imagePixels = convertYUV420_NV21toRGB8888(yuvimage.getYuvData(), previewSize.width, previewSize.height);
        // return Observable.just(Bitmap.createBitmap(imagePixels, previewSize.width, previewSize.height, Bitmap.Config.ARGB_8888));
    }

    companion object {

        // Min allowed camera image size, if it is less setParameters will throw exception
        private const val CAMERA_WIDTH = 1280
        private const val CAMERA_HEIGHT = 720

        /**
         * Converts YUV420 NV21 to RGB8888
         *
         * @param data   byte array on YUV420 NV21 format.
         * @param width  pixels width
         * @param height pixels height
         * @return a RGB8888 pixels int array. Where each int is a pixels ARGB.
         */
        fun convertYUV420_NV21toRGB8888(data: ByteArray, width: Int, height: Int): IntArray {
            val size = width * height
            val pixels = IntArray(size)
            var u: Int
            var v: Int
            var y1: Int
            var y2: Int
            var y3: Int
            var y4: Int

            // i percorre os Y and the final pixels
            // k percorre os pixles U e V
            var i = 0
            var k = 0
            while (i < size) {
                y1 = (data[i] and 0xff.toByte()).toInt()
                y2 = (data[i + 1] and 0xff.toByte()).toInt()
                y3 = (data[width + i] and 0xff.toByte()).toInt()
                y4 = (data[width + i + 1] and 0xff.toByte()).toInt()

                u = (data[size + k] and 0xff.toByte()).toInt()
                v = (data[size + k + 1] and 0xff.toByte()).toInt()
                u -= 128
                v -= 128

                pixels[i] = convertYUVtoRGB(y1, u, v)
                pixels[i + 1] = convertYUVtoRGB(y2, u, v)
                pixels[width + i] = convertYUVtoRGB(y3, u, v)
                pixels[width + i + 1] = convertYUVtoRGB(y4, u, v)

                if (i != 0 && (i + 2) % width == 0)
                    i += width
                i += 2
                k += 2
            }

            return pixels
        }

        private fun convertYUVtoRGB(y: Int, u: Int, v: Int): Int {
            var r = y + (1.402f * v).toInt()
            var g = y - (0.344f * u + 0.714f * v).toInt()
            var b = y + (1.772f * u).toInt()

            r = if (r > 255) 255 else if (r < 0) 0 else r
            g = if (g > 255) 255 else if (g < 0) 0 else g
            b = if (b > 255) 255 else if (b < 0) 0 else b
            return -0x1000000 or (b shl 16) or (g shl 8) or r
        }
    }
}
